src/cmd/ksh93/edit/history.c

```diff
@@ -310,6 +310,8 @@ int  sh_histinit(void *sh_context)
 		sh_close(fd);
 		return(0);
 	}
+
+	memset(hp, 0, histmask*sizeof(off_t));
```

Comment:
Please explain to me what the preceding allocation is doing and why this is correct. The *src/cmd/ksh93/include/defs.h* module says

```
#define new_of(type,x)  ((type*)malloc((unsigned)sizeof(type)+(x)))
```

Note that this module has the same definition guarded by `#if !KSHELL`. No idea what that is all about. But in any case why is it subtracting from the size of the `History_t` struct:

```
if(!(hp=new_of(History_t,(--histmask)*sizeof(off_t))))
```

`histmask` is just a power of two so it shouldn't be negative which means that this use should not be negated AFAICT. There isn't a single other use of `new_of()` that passes a negative value for the second parameter.
So this memset looks correct but the preceding `new_of()` is either wrong or needs a comment explaining what is going on.

Reply:
Also, there aren't many uses of `new_of` and it just obfuscates the code. Seems to me the right thing to do is replace its uses in this function with `malloc()` and just do the addition explicitly rather than hiding it in that `new_of()` macro.

Reply:
Yeah, the code is strange. The for loop tries to get a number that is greater than number of history items to allocate and is a multiple of 2. Not sure why 1 is being subtracted later. We can probably refactor it.

Reply:
The last member of ```History_t``` is ```off_t   histcmds[2]```, which is nullified separately by:
```
memset((char*)hp->histcmds,0,sizeof(off_t)*(hp->histmask+1));
```

If you want to nullify the metadata, it should be written like this:
```
memset(hp, 0, sizeof(History_t));
```
... or just:
```
memset(hp, 0, sizeof *hp);
```
... which is less prone to programming mistakes.

Reply:
```memset(hp, 0, sizeof(History_t));```
```memset(hp, 0, sizeof *hp);```

I don't think we can use ```sizeof``` operator here because the amount of memory allocated to ```hp``` is not ```sizeof(History_t)```. It is calculated separately in the code above in a ```for``` loop.